dataset/components/version/views.py

```diff
@@ -0,0 +1,154 @@
+    version_crud: VersionCRUD = Depends(get_version_crud),
+    version_published: PublishVersion = Depends(get_version_publisher),
+):
+    """Create a version in Minio."""
+    await version_crud.check_duplicate_versions(data.version, dataset_id)
+    dataset = await dataset_crud.retrieve_by_id(dataset_id)
+    background_tasks.add_task(version_published.publish, dataset.code, dataset.id, data)
+    return LegacyResponseSchema(result={'status_id': dataset_id})
+
+
+@router.get(
+    '/v1/dataset/{dataset_id}/publish/status',
+    tags=['version'],
+    summary='Publish status',
+    response_model=LegacyResponseSchema,
+)
+async def publish_status(
+    dataset_id: str,
+    status_id: str,
+    version_crud: VersionCRUD = Depends(get_version_crud),
+    redis_client: StrictRedis = Depends(get_redis_client),
+):
+    """Get status of publish background task."""
+    await version_crud.get_version(dataset_id)
+    status = await redis_client.get(status_id)
+    if not status:
+        raise NotFound()
+    return LegacyResponseSchema(result=json.loads(status))
+
+
+@router.get(
+    '/v1/dataset/versions',
+    tags=['version'],
+    response_model=VersionListResponseSchema,
+    summary='Get dataset versions',
+)
+async def version(
+    page_parameters: PageParameters = Depends(),
+    filter_parameters: VersionFilterParameters = Depends(),
+    sort_parameters: SortParameters.with_sort_by_fields(VersionSortByFields) = Depends(),
+    version_crud: VersionCRUD = Depends(get_version_crud),
+):
+    """Get list of a versions from dataset."""
+    filtering = filter_parameters.to_filtering()
+    sorting = sort_parameters.to_sorting()
+    pagination = page_parameters.to_pagination()
+
+    async with version_crud:
+        page = await version_crud.paginate(pagination, sorting, filtering)
+    response = VersionListResponseSchema.from_page(page)
+
+    return response
+
+
+@router.delete(
+    '/v1/dataset/{dataset_id}/version/{version_id}',
+    tags=['version'],
+    summary='Only used for unit tests, delete a version from psql',
+)
+async def delete_version(
+    dataset_id: str,
+    version_id: int,
+    version_crud: VersionCRUD = Depends(get_version_crud),
+) -> Response:
+    """Delete a version."""
+    await version_crud.delete(version_id)
+    return Response(status_code=204)
+
+
+@router.get(
+    '/v1/dataset/{dataset_id}/download/pre',
+    tags=['version'],
+    summary='Download dataset version',
+    response_model=LegacyResponseSchema,
+)
+async def download_url(
+    dataset_id: str,
+    version: str,
+    version_crud: DatasetCRUD = Depends(get_version_crud),
+    s3_client: S3Client = Depends(get_s3_client),
+):
+    """Get download url for dataset version."""
+    version = await version_crud.get_version(dataset_id, version)
+    # generate the minio presigned url
+    minio_path = version.location.split('//')[-1]
+    _, bucket, file_path = tuple(minio_path.split('/', 2))
```

Comment:
I suggest to move this into a `Version` class `generate_presigned_url` method:
```suggestion
    # generate the minio presigned url
    minio_path = version.location.split('//')[-1]
    _, bucket, file_path = tuple(minio_path.split('/', 2))
```

Reply:
This code is duplicated in the service. I want to revisit this whole file/folder domain in this ticket:
https://indocconsortium.atlassian.net/browse/PILOT-1974